File: sdn-js/src/channels.ts
import { isValidSchema, type SchemaName } from './schemas';

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
const STANDARD_CODE_RE = /^[A-Z]{3}$/;
const INTERNAL_SCHEMA_SUFFIX = String.fromCharCode(46, 102, 98, 115);

export const CHANNEL_TOPIC_PREFIX = '/spacedatanetwork/channels/';

export interface ParsedChannelId {
  channelId: string;
  sourceId: string;
  standardCode: string;
  feedUuid: string | null;
}

export interface ChannelIdInput {
  sourceId: string;
  standardCode: string;
  feedUuid?: string | null;
}

export function assertStandardCode(value: string): string {
  const code = value.trim();
  if (!STANDARD_CODE_RE.test(code)) {
    throw new Error(`invalid standardCode ${JSON.stringify(value)}`);
  }
  if (!isValidSchema(`${code}${INTERNAL_SCHEMA_SUFFIX}` as SchemaName)) {
    throw new Error(`unknown standardCode ${JSON.stringify(value)}`);
  }
  return code;
}

export function schemaNameFromStandardCode(standardCode: string): SchemaName {
  const code = assertStandardCode(standardCode);
  return `${code}${INTERNAL_SCHEMA_SUFFIX}` as SchemaName;
}

export function standardCodeFromSchemaName(schemaName: string): string {
  const value = schemaName.trim();
  if (!value.endsWith(INTERNAL_SCHEMA_SUFFIX)) {
    throw new Error(`invalid schema name ${JSON.stringify(schemaName)}`);
  }
  return assertStandardCode(value.slice(0, -INTERNAL_SCHEMA_SUFFIX.length));
}

export function parseChannelId(channelId: string): ParsedChannelId {
  const value = channelId.trim();
  if (value === '' || value.includes('--')) {
    throw new Error(`invalid channel ID ${JSON.stringify(channelId)}`);
  }

  const parts = value.split('-');
  if (parts.length < 2 || parts.some((part) => part === '')) {
    throw new Error(`invalid channel ID ${JSON.stringify(channelId)}`);
  }

  let feedUuid: string | null = null;
  let standardIndex = parts.length - 1;
  if (parts.length >= 7) {
    const maybeUuid = parts.slice(-5).join('-');
    if (UUID_RE.test(maybeUuid)) {
      feedUuid = maybeUuid;
      standardIndex = parts.length - 6;
    }
  }

  if (standardIndex <= 0) {
    throw new Error(`channel sourceId is required in ${JSON.stringify(channelId)}`);
  }

  const standardCode = assertStandardCode(parts[standardIndex]);
  const sourceId = parts.slice(0, standardIndex).join('-');
  if (sourceId === '') {
    throw new Error(`channel sourceId is required in ${JSON.stringify(channelId)}`);
  }

  return {
    channelId: formatChannelId({ sourceId, standardCode, feedUuid }),
    sourceId,
    standardCode,
    feedUuid,
  };
}

export function formatChannelId(input: ChannelIdInput): string {
  const sourceId = input.sourceId.trim();
  if (sourceId === '' || sourceId.includes('--')) {
    throw new Error('channel sourceId is required');
  }

  const standardCode = assertStandardCode(input.standardCode);
  const feedUuid = input.feedUuid == null ? null : input.feedUuid.trim();
  if (feedUuid && !UUID_RE.test(feedUuid)) {
    throw new Error(`invalid feedUuid ${JSON.stringify(input.feedUuid)}`);
  }

  return feedUuid ? `${sourceId}-${standardCode}-${feedUuid}` : `${sourceId}-${standardCode}`;
}

export function channelDiscoveryTopic(standardCode: string): string {
  return `${CHANNEL_TOPIC_PREFIX}${assertStandardCode(standardCode)}`;
}

File: sdn-server/internal/channels/channels.go
// Package channels defines the public SDN channel naming contract.
package channels

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/spacedatanetwork/sdn-server/internal/sds"
)

const DiscoveryTopicPrefix = "/spacedatanetwork/channels/"

var (
	uuidPattern          = regexp.MustCompile(`(?i)^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`)
	standardCodePattern  = regexp.MustCompile(`^[A-Z]{3}$`)
	internalSchemaSuffix = string([]byte{'.', 'f', 'b', 's'})
)

type ChannelID struct {
	ChannelID    string
	SourceID     string
	StandardCode string
	FeedUUID     string
}

type ChannelIDInput struct {
	SourceID     string
	StandardCode string
	FeedUUID     string
}

func ParseChannelID(channelID string) (ChannelID, error) {
	value := strings.TrimSpace(channelID)
	if value == "" || strings.Contains(value, "--") {
		return ChannelID{}, fmt.Errorf("invalid channel ID %q", channelID)
	}
	parts := strings.Split(value, "-")
	if len(parts) < 2 {
		return ChannelID{}, fmt.Errorf("invalid channel ID %q", channelID)
	}
	for _, part := range parts {
		if part == "" {
			return ChannelID{}, fmt.Errorf("invalid channel ID %q", channelID)
		}
	}

	standardIndex := len(parts) - 1
	feedUUID := ""
	if len(parts) >= 7 {
		maybeUUID := strings.Join(parts[len(parts)-5:], "-")
		if uuidPattern.MatchString(maybeUUID) {
			feedUUID = maybeUUID
			standardIndex = len(parts) - 6
		}
	}
	if standardIndex <= 0 {
		return ChannelID{}, fmt.Errorf("channel sourceId is required in %q", channelID)
	}

	standardCode, err := AssertStandardCode(parts[standardIndex])
	if err != nil {
		return ChannelID{}, err
	}
	sourceID := strings.Join(parts[:standardIndex], "-")
	formatted, err := FormatChannelID(ChannelIDInput{
		SourceID:     sourceID,
		StandardCode: standardCode,
		FeedUUID:     feedUUID,
	})
	if err != nil {
		return ChannelID{}, err
	}
	return ChannelID{
		ChannelID:    formatted,
		SourceID:     sourceID,
		StandardCode: standardCode,
		FeedUUID:     feedUUID,
	}, nil
}

func FormatChannelID(input ChannelIDInput) (string, error) {
	sourceID := strings.TrimSpace(input.SourceID)
	if sourceID == "" || strings.Contains(sourceID, "--") {
		return "", fmt.Errorf("channel sourceId is required")
	}
	standardCode, err := AssertStandardCode(input.StandardCode)
	if err != nil {
		return "", err
	}
	feedUUID := strings.TrimSpace(input.FeedUUID)
	if feedUUID != "" && !uuidPattern.MatchString(feedUUID) {
		return "", fmt.Errorf("invalid feedUuid %q", input.FeedUUID)
	}
	if feedUUID != "" {
		return sourceID + "-" + standardCode + "-" + feedUUID, nil
	}
	return sourceID + "-" + standardCode, nil
}

func AssertStandardCode(value string) (string, error) {
	code := strings.TrimSpace(value)
	if !standardCodePattern.MatchString(code) {
		return "", fmt.Errorf("invalid standardCode %q", value)
	}
	if err := sds.ValidateSchemaName(code + internalSchemaSuffix); err != nil {
		return "", fmt.Errorf("invalid standardCode %q: %w", value, err)
	}
	for _, schemaName := range sds.SupportedSchemas {
		if schemaName == code+internalSchemaSuffix {
			return code, nil
		}
	}
	if sds.IsPublishedBindingSchema(code + internalSchemaSuffix) {
		return code, nil
	}
	return "", fmt.Errorf("unknown standardCode %q", value)
}

func SchemaNameFromStandardCode(standardCode string) (string, error) {
	code, err := AssertStandardCode(standardCode)
	if err != nil {
		return "", err
	}
	return code + internalSchemaSuffix, nil
}

func StandardCodeFromSchemaName(schemaName string) (string, error) {
	value := strings.TrimSpace(schemaName)
	if !strings.HasSuffix(value, internalSchemaSuffix) {
		return "", fmt.Errorf("invalid schema name %q", schemaName)
	}
	return AssertStandardCode(strings.TrimSuffix(value, internalSchemaSuffix))
}

func DiscoveryTopic(standardCode string) string {
	code, err := AssertStandardCode(standardCode)
	if err != nil {
		return ""
	}
	return DiscoveryTopicPrefix + code
}
